### ISSUE TITLE
Feature Request/Suggestion: Ability to `.reshape(...)` array when chunk size unknown

### ISSUE THREAD

**Issue by asmith26** [dask/dask#4469]
https://github.com/dask/dask/issues/1839 discusses about having the ability to `.reshape(...)` a dask array when the chunk size is unknown.

I am converting dataframe -> array, and would like to be able to reshape it.

One idea I had is add the ability to pass in a "length" or "chunk_size" argument to [`df.to_records()`](http://docs.dask.org/en/latest/array-creation.html#from-dask-dataframe), since for my particular problem I know this.

Would this be of interest? I would be happy to have a go at an implementation if so.

**Comment by mrocklin** [comment]
I believe that the `DataFrame.to_dask_array` method already supports a `lengths=` parameter, so there is good precedence for this (and a good API and implementation to follow).  Adding that also to `to_records` sounds good to me.

**Comment by asmith26** [comment]
Referencing for infomation: https://github.com/dask/dask/issues/3090

**Comment by mrocklin** [comment]
@asmith26 do you have any interest in contributing and doing this work?

**Comment by asmith26** [comment]
Hi @mrocklin, yep I'm very interested in contributing (or attempting to contribute) to this fantastic library.

Please find my first commit at: https://github.com/asmith26/dask/commit/560722dba9ba394d614a7486ebf6e623757575b5

Notes:
- This is my first commit to dask, so any feedback/suggestions welcome.
- I've heavily followed the[ `.to_dask_array` code](https://github.com/asmith26/dask/blob/master/dask/dataframe/core.py#L1105-L1150) (and corresponding [test](https://github.com/asmith26/dask/blob/master/dask/dataframe/tests/test_dataframe.py#L2043-L2063))
- Unfortunately by copying this code, in the case when `lengths=[2, 3]` I get:
```
F
dask/dataframe/io/tests/test_io.py:498 (test_to_records[lengths1])
lengths = [2, 3]

    @pytest.mark.parametrize('lengths', [
        None,
        [2, 3],
        True,
    ])
    def test_to_records(lengths):
        pytest.importorskip('dask.array')
        from dask.array.utils import assert_eq
        df = pd.DataFrame({'x': ['a', 'b', 'c', 'd'],
                           'y': [2, 3, 4, 5]},
                          index=pd.Index([1., 2., 3., 4.], name='ind'))
        ddf = dd.from_pandas(df, 2)
    
>       assert_eq(df.to_records(), ddf.to_records(lengths=lengths))

test_io.py:512: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../array/utils.py:103: in assert_eq
    b = b.compute(scheduler='sync')
../../../base.py:156: in compute
    (result,) = compute(self, traverse=False, **kwargs)
../../../base.py:395: in compute
    dsk = collections_to_dsk(collections, optimize_graph, **kwargs)
../../../base.py:194: in collections_to_dsk
    for opt, (dsk, keys) in groups.items()]))
../../../base.py:194: in <listcomp>
    for opt, (dsk, keys) in groups.items()]))
../../../array/optimization.py:45: in optimize
    dsk2, dependencies = cull(dsk, keys)
../../../optimization.py:46: in cull
    for k in work]
../../../optimization.py:46: in <listcomp>
    for k in work]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

dsk = {('from_pandas-b581c7965fed6d41f34e734ae95a309c', 0):      x  y
ind      
1.0  a  2
2.0  b  3, ('from_pandas-b581c7965...ubgraph_callable, ('from_pandas-b581c7965fed6d41f34e734ae95a309c', 1), 'from_pandas-b581c7965fed6d41f34e734ae95a309c')}
key = ('to_records-7efaccf97358a6d0b69926d9bf624a17', 1, 0), task = None
as_list = True

    def get_dependencies(dsk, key=None, task=None, as_list=False):
        """ Get the immediate tasks on which this task depends
    
        Examples
        --------
        >>> dsk = {'x': 1,
        ...        'y': (inc, 'x'),
        ...        'z': (add, 'x', 'y'),
        ...        'w': (inc, 'z'),
        ...        'a': (add, (inc, 'x'), 1)}
    
        >>> get_dependencies(dsk, 'x')
        set()
    
        >>> get_dependencies(dsk, 'y')
        {'x'}
    
        >>> get_dependencies(dsk, 'z')  # doctest: +SKIP
        {'x', 'y'}
    
        >>> get_dependencies(dsk, 'w')  # Only direct dependencies
        {'z'}
    
        >>> get_dependencies(dsk, 'a')  # Ignore non-keys
        {'x'}
    
        >>> get_dependencies(dsk, task=(inc, 'x'))  # provide tasks directly
        {'x'}
        """
        if key is not None:
>           arg = dsk[key]
E           KeyError: ('to_records-7efaccf97358a6d0b69926d9bf624a17', 1, 0)

../../../core.py:187: KeyError
```

Any idea? The main difference for `to_records` is that it refers to many other functions called `to_records`:
- From the main `to_records` function I've been working on, it goes to [ this `to_records`](https://github.com/asmith26/dask/blob/master/dask/dataframe/io/io.py#L469-L486)
- Then I'm lost: does `M.to_records` refer to the pandas function maybe?

**Comment by mrocklin** [comment]
@asmith26 I recommend including your commit as a WIP pull request.  It will be easier to have conversation there.

**Comment by asmith26** [comment]
Thanks for your help. Please find WIP pull request at https://github.com/dask/dask/pull/4515

**Comment by jrbourbeau** [comment]
Closed via #4515
